client/client_auth_test.go

```diff
@@ -164,12 +167,67 @@ func TestClientAuthUnknownAuthorityCert(t *testing.T) {
 	rsp, err := client.Request(ac, ts.URL, msger)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "certificate signed by unknown authority")
+	// see https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/x509/x509_vfy.c#L3268
+	//     for self-signed openssl always returns self-signed error; either
+	//     X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT or X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN
+	//     and never X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT or X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY
+	assert.Nil(t, rsp)
+}
+
+//X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT
+func TestClientAuthDepthZeroSelfSignedCert(t *testing.T) {
+	ts := startTestHTTPS(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}),
+		localhostCert,
+		localhostKey)
+	defer ts.Close()
+
+	ac, err := NewApiClient(
+		Config{"server.zero.depth.self.signed.crt", true, false},
+	)
+	assert.NotNil(t, ac)
+	assert.NoError(t, err)
+
+	client := NewAuth()
+	assert.NotNil(t, client)
+
+	msger := &testAuthDataMessenger{
+		reqData: []byte("foobar"),
+	}
+	rsp, err := client.Request(ac, ts.URL, msger)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "depth zero self-signed certificate")
+	assert.Nil(t, rsp)
+}
+
+//X509_V_ERR_EE_KEY_TOO_SMALL
+func TestClientAuthEndEntityKeyTooSmall(t *testing.T) {
+	ts := startTestHTTPS(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}),
+		localhostCertShortEEKey,
+		localhostKeyShortEEKey)
+	defer ts.Close()
+
+	ac, err := NewApiClient(
+		Config{"server.crt", true, false},
+	)
+	assert.NotNil(t, ac)
+	assert.NoError(t, err)
+
+	client := NewAuth()
+	assert.NotNil(t, client)
+
+	msger := &testAuthDataMessenger{
+		reqData: []byte("foobar"),
+	}
+	rsp, err := client.Request(ac, ts.URL, msger)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "end entity key too short")
```

Comment:
I was curious and tried to run these tests locally. However, this test failed with the following error:
```
    client_auth_test.go:223: 
        	Error Trace:	client_auth_test.go:223
        	Error:      	"generic error occurred while executing authorization request: Post \"https://127.0.0.1:41371/api/devices/v1/authentication/auth_requests\": depth zero self-signed certificate, openssl verify rc: 18 server cert file: server.crt" does not contain "end entity key too short"
        	Test:       	TestClientAuthEndEntityKeyTooSmall
--- FAIL: TestClientAuthEndEntityKeyTooSmall (0.00s)
```

Reply:
what s your security level? `cat /etc/ssl/openssl.cnf` please?

Reply:
IMHO, I think the implementation should be resilient to openssl configuration if possible.

Reply:
I would merge this, and then create follow up, without [this one](https://github.com/spacemonkeygo/openssl/pull/140) we will have to fork openssl wrapper. We might do it anyways, but for the sake of continuing, I would beg to try to make it work at level2 and then first thing next free moment do as you say. since I kind of agree :)